google/cloud/internal/win32/oauth2_service_account_credentials.cc

```diff
@@ -0,0 +1,219 @@
+  if (!CryptEncodeObjectEx(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                           PKCS_RSA_PRIVATE_KEY, exported_key.data(), 0,
+                           nullptr, nullptr, &pkcs1_encoded_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("Could not encode private key from PKCS#12 file (",
+                          source, "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<BYTE> pkcs1_encoded(pkcs1_encoded_length);
+  CryptEncodeObjectEx(X509_ASN_ENCODING | PKCS_7_ASN_ENCODING,
+                      PKCS_RSA_PRIVATE_KEY, exported_key.data(), 0, nullptr,
+                      pkcs1_encoded.data(), &pkcs1_encoded_length);
+
+  // Wrap the PKCS#1 encoded private key in a PKCS#8 structure.
+  CRYPT_PRIVATE_KEY_INFO private_key_info;
+  private_key_info.Version = 0;
+  private_key_info.Algorithm.pszObjId = szOID_RSA_RSA;
+  private_key_info.Algorithm.Parameters.cbData = 0;
+  private_key_info.Algorithm.Parameters.pbData = nullptr;
+  private_key_info.PrivateKey.cbData = pkcs1_encoded_length;
+  private_key_info.PrivateKey.pbData = pkcs1_encoded.data();
+  private_key_info.pAttributes = nullptr;
+  DWORD pkcs8_encoded_length;
+  if (!CryptEncodeObjectEx(X509_ASN_ENCODING, PKCS_PRIVATE_KEY_INFO,
+                           &private_key_info, 0, nullptr, nullptr,
+                           &pkcs8_encoded_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors("Could not encode private key from PKCS#12 file (",
+                          source, "): "),
+        GCP_ERROR_INFO());
+  }
+  std::vector<BYTE> pkcs8_encoded(pkcs8_encoded_length);
+  CryptEncodeObjectEx(X509_ASN_ENCODING, PKCS_PRIVATE_KEY_INFO,
+                      &private_key_info, 0, nullptr, pkcs8_encoded.data(),
+                      &pkcs8_encoded_length);
+
+  // Convert to base64 and add the PEM markers.
+  DWORD base64_length;
+  if (!CryptBinaryToStringA(pkcs8_encoded.data(), pkcs8_encoded_length,
+                            CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR, nullptr,
+                            &base64_length)) {
+    return InvalidArgumentError(
+        FormatWin32Errors(
+            "Could not base64 encode private key from PKCS#12 file (", source,
+            "): "),
+        GCP_ERROR_INFO());
+  }
+  std::string private_key = "-----BEGIN PRIVATE KEY-----\n";
+  // Enlarge the string and directly write the base64 data into it.
+  private_key.resize(private_key.size() + base64_length);
+  CryptBinaryToStringA(pkcs8_encoded.data(), pkcs8_encoded_length,
+                       CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR,
+                       &*(private_key.end() - base64_length), &base64_length);
+  private_key.pop_back();  // remove the null terminator
+  private_key += "-----END PRIVATE KEY-----\n";
+
+  return ServiceAccountCredentialsInfo{std::move(service_account_id),
+                                       kP12PrivateKeyIdMarker,
+                                       std::move(private_key),
```

Comment:
Maybe this is more readable:

```suggestion
  std::vector<char> private_key(base64_length);
  CryptBinaryToStringA(pkcs8_encoded.data(), pkcs8_encoded_length,
                       CRYPT_STRING_BASE64 | CRYPT_STRING_NOCR,
                       private_key.data(), &base64_length);
  private_key.pop_back();  // remove the null terminator

    return ServiceAccountCredentialsInfo{std::move(service_account_id),
                                       kP12PrivateKeyIdMarker,
                                       absl::StrCat("-----BEGIN PRIVATE KEY-----\n", absl::string_view(private_key.data(), private_key.size()), "-----END PRIVATE KEY-----\n"),
```

Reply:
Done. This could have been using our own Base64 code but let's not do it in this PR.